File: cpp/src/Ice/MetricsAdminI.cpp
// Copyright (c) ZeroC, Inc.

#include "MetricsAdminI.h"
#include "Ice/Logger.h"
#include "Ice/LoggerUtil.h"
#include "Ice/Properties.h"
#include "Ice/StringUtil.h"
#include "MetricsObserverI.h"

#include <chrono>
#include <stdexcept>

using namespace std;
using namespace Ice;
using namespace IceInternal;
using namespace IceMX;

namespace
{
    // NOLINTBEGIN(cert-err58-cpp)
    const string suffixes[] = {
        "Disabled",
        "GroupBy",
        "Accept.*",
        "Reject.*",
        "RetainDetached",
        "Map.*",
    };
    // NOLINTEND(cert-err58-cpp)

    void validateProperties(const string& prefix, const PropertiesPtr& properties)
    {
        vector<string> unknownProps;
        PropertyDict props = properties->getPropertiesForPrefix(prefix);
        for (const auto& p : props)
        {
            bool valid = false;
            for (const auto& suffix : suffixes)
            {
                string prop = prefix + suffix;
                if (IceInternal::match(p.first, prop))
                {
                    valid = true;
                    break;
                }
            }
            if (!valid)
            {
                unknownProps.push_back(p.first);
            }
        }

        if (!unknownProps.empty())
        {
            ostringstream os;
            os << "found unknown properties for " << "IceMX" << ": '" << prefix << "'";
            for (const auto& prop : unknownProps)
            {
                os << "\n    " << prop;
            }
            throw PropertyException{__FILE__, __LINE__, os.str()};
        }
    }

    vector<MetricsMapI::RegExpPtr> parseRule(const PropertiesPtr& properties, const string& name)
    {
        vector<MetricsMapI::RegExpPtr> regexps;
        PropertyDict rules = properties->getPropertiesForPrefix(name + '.');
        for (const auto& rule : rules)
        {
            try
            {
                regexps.push_back(make_shared<MetricsMapI::RegExp>(rule.first.substr(name.length() + 1), rule.second));
            }
            catch (const std::exception&)
            {
                throw invalid_argument("invalid regular expression '" + rule.second + "' for '" + rule.first + "'");
            }
        }
        return regexps;
    }
}

MetricsMapI::RegExp::RegExp(string attribute, const string& regexp) : _attribute(std::move(attribute))
{
    _regex = regex(regexp, std::regex_constants::extended | std::regex_constants::nosubs);
}

bool
MetricsMapI::RegExp::match(const string& value)
{
    return regex_match(value, _regex);
}

MetricsMapI::~MetricsMapI() = default; // Out of line to avoid weak vtable

MetricsMapI::MetricsMapI(const std::string& mapPrefix, const PropertiesPtr& properties)
    : _properties(properties->getPropertiesForPrefix(mapPrefix)),
      _retain(properties->getIcePropertyAsInt(mapPrefix + "RetainDetached")),
      _accept(parseRule(properties, mapPrefix + "Accept")),
      _reject(parseRule(properties, mapPrefix + "Reject"))
{
    validateProperties(mapPrefix, properties);

    string groupBy = properties->getPropertyWithDefault(mapPrefix + "GroupBy", "id");
    auto& groupByAttributes = const_cast<vector<string>&>(_groupByAttributes);
    auto& groupBySeparators = const_cast<vector<string>&>(_groupBySeparators);
    if (!groupBy.empty())
    {
        string v;
        bool attribute = IceInternal::isAlpha(groupBy[0]) || IceInternal::isDigit(groupBy[0]);
        if (!attribute)
        {
            groupByAttributes.emplace_back("");
        }

        for (char p : groupBy)
        {
            bool isAlphaNum = IceInternal::isAlpha(p) || IceInternal::isDigit(p) || p == '.';
            if (attribute && !isAlphaNum)
            {
                groupByAttributes.push_back(v);
                v = p;
                attribute = false;
            }
            else if (!attribute && isAlphaNum)
            {
                groupBySeparators.push_back(v);
                v = p;
                attribute = true;
            }
            else
            {
                v += p;
            }
        }

        if (attribute)
        {
            groupByAttributes.push_back(v);
        }
        else
        {
            groupBySeparators.push_back(v);
        }
    }
}

MetricsMapI::MetricsMapI(const MetricsMapI& map)
    : std::enable_shared_from_this<MetricsMapI>(),
      _properties(map._properties),
      _groupByAttributes(map._groupByAttributes),
      _groupBySeparators(map._groupBySeparators),
      _retain(map._retain),
      _accept(map._accept),
      _reject(map._reject)
{
}

const Ice::PropertyDict&
MetricsMapI::getProperties() const
{
    return _properties;
}

MetricsMapFactory::~MetricsMapFactory() = default; // Out of line to avoid weak vtable

MetricsMapFactory::MetricsMapFactory(Updater* updater) : _updater(updater) {}

void
MetricsMapFactory::update()
{
    assert(_updater);
    _updater->update();
}

MetricsViewI::MetricsViewI(string name) : _name(std::move(name)) {}

void
MetricsViewI::destroy()
{
    for (const auto& map : _maps)
    {
        map.second->destroy();
    }
}

bool
MetricsViewI::addOrUpdateMap(
    const PropertiesPtr& properties,
    const string& mapName,
    const MetricsMapFactoryPtr& factory,
    const Ice::LoggerPtr& logger)
{
    const string viewPrefix = "IceMX.Metrics." + _name + ".";
    const string mapsPrefix = viewPrefix + "Map.";
    PropertyDict mapsProps = properties->getPropertiesForPrefix(mapsPrefix);

    string mapPrefix;
    PropertyDict mapProps;
    if (!mapsProps.empty())
    {
        mapPrefix = mapsPrefix + mapName + ".";
        mapProps = properties->getPropertiesForPrefix(mapPrefix);
        if (mapProps.empty())
        {
            // This map isn't configured for this view.
            auto q = _maps.find(mapName);
            if (q != _maps.end())
            {
                q->second->destroy();
                _maps.erase(q);
                return true;
            }
            return false;
        }
    }
    else
    {
        mapPrefix = viewPrefix;
        mapProps = properties->getPropertiesForPrefix(mapPrefix);
    }

    if (properties->getPropertyAsInt(mapPrefix + "Disabled") > 0)
    {
        // This map is disabled for this view.
        auto q = _maps.find(mapName);
        if (q != _maps.end())
        {
            q->second->destroy();
            _maps.erase(q);
            return true;
        }
        return false;
    }

    auto q = _maps.find(mapName);
    if (q != _maps.end() && q->second->getProperties() == mapProps)
    {
        return false; // The map configuration didn't change, no need to re-create.
    }

    if (q != _maps.end())
    {
        // Destroy the previous map
        q->second->destroy();
        _maps.erase(q);
    }

    try
    {
        _maps.insert(make_pair(mapName, factory->create(mapPrefix, properties)));
    }
    catch (const std::exception& ex)
    {
        Ice::Warning warn(logger);
        warn << "unexpected exception while creating metrics map:\n" << ex;
    }
    return true;
}

bool
MetricsViewI::removeMap(const string& mapName)
{
    auto q = _maps.find(mapName);
    if (q != _maps.end())
    {
        q->second->destroy();
        _maps.erase(q);
        return true;
    }
    return false;
}

MetricsView
MetricsViewI::getMetrics()
{
    MetricsView metrics;
    for (const auto& map : _maps)
    {
        metrics.insert(make_pair(map.first, map.second->getMetrics()));
    }
    return metrics;
}

MetricsFailuresSeq
MetricsViewI::getFailures(const string& mapName)
{
    auto p = _maps.find(mapName);
    if (p != _maps.end())
    {
        return p->second->getFailures();
    }
    return {};
}

MetricsFailures
MetricsViewI::getFailures(const string& mapName, const string& id)
{
    auto p = _maps.find(mapName);
    if (p != _maps.end())
    {
        return p->second->getFailures(id);
    }
    return {};
}

vector<string>
MetricsViewI::getMaps() const
{
    vector<string> maps;
    maps.reserve(_maps.size());
    for (const auto& map : _maps)
    {
        maps.push_back(map.first);
    }
    return maps;
}

MetricsMapIPtr
MetricsViewI::getMap(const string& mapName) const
{
    auto p = _maps.find(mapName);
    if (p != _maps.end())
    {
        return p->second;
    }
    return nullptr;
}

MetricsAdminI::MetricsAdminI(PropertiesPtr properties, LoggerPtr logger)
    : _logger(std::move(logger)),
      _properties(std::move(properties))
{
    updateViews();
}

MetricsAdminI::~MetricsAdminI() = default;

void
MetricsAdminI::destroy()
{
    lock_guard lock(_mutex);
    for (const auto& view : _views)
    {
        view.second->destroy();
    }
}

void
MetricsAdminI::updateViews()
{
    set<MetricsMapFactoryPtr> updatedMaps;
    {
        lock_guard lock(_mutex);
        const string viewsPrefix = "IceMX.Metrics.";
        PropertyDict viewsProps = _properties->getPropertiesForPrefix(viewsPrefix);
        map<string, MetricsViewIPtr> views;
        _disabledViews.clear();
        for (const auto& viewsProp : viewsProps)
        {
            string viewName = viewsProp.first.substr(viewsPrefix.size());
            string::size_type dotPos = viewName.find('.');
            if (dotPos != string::npos)
            {
                viewName = viewName.substr(0, dotPos);
            }

            if (views.find(viewName) != views.end() || _disabledViews.find(viewName) != _disabledViews.end())
            {
                continue; // View already configured.
            }

            validateProperties(viewsPrefix + viewName + ".", _properties);

            if (_properties->getPropertyAsIntWithDefault(viewsPrefix + viewName + ".Disabled", 0) > 0)
            {
                _disabledViews.insert(viewName);
                continue; // The view is disabled
            }

            //
            // Create the view or update it.
            //
            auto q = _views.find(viewName);
            if (q == _views.end())
            {
                q = views
                        .insert(map<string, MetricsViewIPtr>::value_type(viewName, make_shared<MetricsViewI>(viewName)))
                        .first;
            }
            else
            {
                q = views.insert(make_pair(viewName, q->second)).first;
            }

            for (const auto& fact : _factories)
            {
                if (q->second->addOrUpdateMap(_properties, fact.first, fact.second, _logger))
                {
                    updatedMaps.insert(fact.second);
                }
            }
        }
        _views.swap(views);

        //
        // Go through removed views to collect maps to update.
        //
        for (const auto& view : views)
        {
            if (_views.find(view.first) == _views.end())
            {
                vector<string> maps = view.second->getMaps();
                for (const auto& map : maps)
                {
                    updatedMaps.insert(_factories[map]);
                }
                view.second->destroy();
            }
        }
    }

    //
    // Call the updaters to update the maps.
    //
    for (const auto& updatedMap : updatedMaps)
    {
        updatedMap->update();
    }
}

void
MetricsAdminI::unregisterMap(const std::string& mapName)
{
    bool updated;
    MetricsMapFactoryPtr factory;
    {
        lock_guard lock(_mutex);
        auto p = _factories.find(mapName);
        if (p == _factories.end())
        {
            return;
        }
        factory = p->second;
        _factories.erase(p);
        updated = removeMap(mapName);
    }
    if (updated)
    {
        factory->update();
    }
}

Ice::StringSeq
MetricsAdminI::getMetricsViewNames(Ice::StringSeq& disabledViews, const Current&)
{
    Ice::StringSeq enabledViews;

    lock_guard lock(_mutex);
    for (const auto& view : _views)
    {
        enabledViews.push_back(view.first);
    }
    disabledViews.insert(disabledViews.end(), _disabledViews.begin(), _disabledViews.end());
    return enabledViews;
}

void
MetricsAdminI::enableMetricsView(string viewName, const Current&)
{
    {
        lock_guard lock(_mutex);
        getMetricsView(viewName); // Throws if unknown metrics view.
        _properties->setProperty("IceMX.Metrics." + viewName + ".Disabled", "0");
    }
    updateViews();
}

void
MetricsAdminI::disableMetricsView(string viewName, const Current&)
{
    {
        lock_guard lock(_mutex);
        getMetricsView(viewName); // Throws if unknown metrics view.
        _properties->setProperty("IceMX.Metrics." + viewName + ".Disabled", "1");
    }
    updateViews();
}

MetricsView
MetricsAdminI::getMetricsView(string viewName, int64_t& timestamp, const Current&)
{
    lock_guard lock(_mutex);
    MetricsViewIPtr view = getMetricsView(viewName);
    timestamp =
        chrono::duration_cast<chrono::milliseconds>(std::chrono::system_clock::now().time_since_epoch()).count();
    if (view)
    {
        return view->getMetrics();
    }
    return {};
}

MetricsFailuresSeq
MetricsAdminI::getMapMetricsFailures(string viewName, string map, const Current&)
{
    lock_guard lock(_mutex);
    MetricsViewIPtr view = getMetricsView(viewName);
    if (view)
    {
        return view->getFailures(map);
    }
    return {};
}

MetricsFailures
MetricsAdminI::getMetricsFailures(string viewName, string map, string id, const Current&)
{
    lock_guard lock(_mutex);
    MetricsViewIPtr view = getMetricsView(viewName);
    if (view)
    {
        return view->getFailures(map, id);
    }
    return {};
}

vector<MetricsMapIPtr>
MetricsAdminI::getMaps(const string& mapName) const
{
    lock_guard lock(_mutex);
    vector<MetricsMapIPtr> maps;
    for (const auto& view : _views)
    {
        MetricsMapIPtr map = view.second->getMap(mapName);
        if (map)
        {
            maps.push_back(map);
        }
    }
    return maps;
}

const LoggerPtr&
MetricsAdminI::getLogger() const
{
    return _logger;
}

MetricsViewIPtr
MetricsAdminI::getMetricsView(const std::string& name)
{
    auto p = _views.find(name);
    if (p == _views.end())
    {
        if (_disabledViews.find(name) == _disabledViews.end())
        {
            throw UnknownMetricsView();
        }
        return nullptr;
    }
    return p->second;
}

void
MetricsAdminI::updated(const PropertyDict& props)
{
    for (const auto& prop : props)
    {
        if (prop.first.find("IceMX.") == 0)
        {
            // Update the metrics views using the new configuration.
            try
            {
                updateViews();
            }
            catch (const std::exception& ex)
            {
                Ice::Warning warn(_logger);
                warn << "unexpected exception while updating metrics view configuration:\n" << ex.what();
            }
            return;
        }
    }
}

bool
MetricsAdminI::addOrUpdateMap(const std::string& mapName, const MetricsMapFactoryPtr& factory)
{
    bool updated = false;
    for (const auto& view : _views)
    {
        updated |= view.second->addOrUpdateMap(_properties, mapName, factory, _logger);
    }
    return updated;
}

bool
MetricsAdminI::removeMap(const std::string& mapName)
{
    bool updated = false;
    for (const auto& view : _views)
    {
        updated |= view.second->removeMap(mapName);
    }
    return updated;
}

File: java/src/com.zeroc.ice/src/main/java/com/zeroc/Ice/MetricsAdminI.java
// Copyright (c) ZeroC, Inc.

package com.zeroc.Ice;

import com.zeroc.IceMX.Metrics;
import com.zeroc.IceMX.MetricsAdmin;
import com.zeroc.IceMX.MetricsFailures;
import com.zeroc.IceMX.UnknownMetricsView;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.function.Consumer;

/**
 * @hidden Public because it's used by IceMX.
 */
public class MetricsAdminI
    implements MetricsAdmin, Consumer<Map<String, String>> {
    private static final String[] suffixes = {
        "Disabled", "GroupBy", "Accept.*", "Reject.*", "RetainDetached", "Map.*",
    };

    static void validateProperties(String prefix, Properties properties) {
        Map<String, String> props = properties.getPropertiesForPrefix(prefix);
        List<String> unknownProps = new ArrayList<>();
        for (String prop : props.keySet()) {
            boolean valid = false;
            for (String suffix : suffixes) {
                if (StringUtil.match(prop, prefix + suffix, false)) {
                    valid = true;
                    break;
                }
            }

            if (!valid) {
                unknownProps.add(prop);
            }
        }

        if (unknownProps.size() > 0) {
            throw new PropertyException(
                "found unknown IceMX properties for:"
                    + ": '"
                    + prefix
                    + "'\n    "
                    + String.join("\n    ", unknownProps));
        }
    }

    static class MetricsMapFactory<T extends Metrics> {
        public MetricsMapFactory(Runnable updater, Class<T> cl) {
            _updater = updater;
            _class = cl;
        }

        public void update() {
            assert (_updater != null);
            _updater.run();
        }

        public MetricsMap<T> create(String mapPrefix, Properties properties) {
            return new MetricsMap<T>(mapPrefix, _class, properties, _subMaps);
        }

        public <S extends Metrics> void registerSubMap(
                String subMap, Class<S> cl, java.lang.reflect.Field field) {
            _subMaps.put(subMap, new MetricsMap.SubMapFactory<S>(cl, field));
        }

        private final Runnable _updater;
        private final Class<T> _class;
        private final Map<String, MetricsMap.SubMapFactory<?>> _subMaps =
            new HashMap<>();
    }

    public MetricsAdminI(Properties properties, Logger logger) {
        _logger = logger;
        _properties = properties;
        updateViews();
    }

    public void updateViews() {
        Set<MetricsMapFactory<?>> updatedMaps = new HashSet<>();
        synchronized (this) {
            String viewsPrefix = "IceMX.Metrics.";
            Map<String, String> viewsProps =
                _properties.getPropertiesForPrefix(viewsPrefix);
            Map<String, MetricsViewI> views = new HashMap<>();
            _disabledViews.clear();
            for (Map.Entry<String, String> e : viewsProps.entrySet()) {
                String viewName = e.getKey().substring(viewsPrefix.length());
                int dotPos = viewName.indexOf('.');
                if (dotPos > 0) {
                    viewName = viewName.substring(0, dotPos);
                }

                if (views.containsKey(viewName) || _disabledViews.contains(viewName)) {
                    continue; // View already configured.
                }

                validateProperties(viewsPrefix + viewName + '.', _properties);

                if (_properties.getPropertyAsIntWithDefault(viewsPrefix + viewName + ".Disabled", 0)
                    > 0) {
                    _disabledViews.add(viewName);
                    continue; // The view is disabled
                }

                //
                // Create the view or update it.
                //
                MetricsViewI v = _views.get(viewName);
                if (v == null) {
                    v = new MetricsViewI(viewName);
                }
                views.put(viewName, v);

                for (Map.Entry<String, MetricsMapFactory<?>> f : _factories.entrySet()) {
                    if (v.addOrUpdateMap(_properties, f.getKey(), f.getValue(), _logger)) {
                        updatedMaps.add(f.getValue());
                    }
                }
            }
            Map<String, MetricsViewI> tmp = _views;
            _views = views;
            views = tmp;

            //
            // Go through removed views to collect maps to update.
            //
            for (Map.Entry<String, MetricsViewI> v : views.entrySet()) {
                if (!_views.containsKey(v.getKey())) {
                    for (String n : v.getValue().getMaps()) {
                        updatedMaps.add(_factories.get(n));
                    }
                }
            }
        }

        //
        // Call the updaters to update the maps.
        //
        for (MetricsMapFactory<?> f : updatedMaps) {
            f.update();
        }
    }

    @Override
    public synchronized MetricsAdmin.GetMetricsViewNamesResult getMetricsViewNames(
            Current current) {
        MetricsAdmin.GetMetricsViewNamesResult r = new MetricsAdmin.GetMetricsViewNamesResult();
        r.disabledViews = _disabledViews.toArray(new String[_disabledViews.size()]);
        r.returnValue = _views.keySet().toArray(new String[_views.size()]);
        return r;
    }

    @Override
    public void enableMetricsView(String name, Current current) throws UnknownMetricsView {
        synchronized (this) {
            getMetricsView(name); // Throws if unknown view.
            _properties.setProperty("IceMX.Metrics." + name + ".Disabled", "0");
        }
        updateViews();
    }

    @Override
    public void disableMetricsView(String name, Current current) throws UnknownMetricsView {
        synchronized (this) {
            getMetricsView(name); // Throws if unknown view.
            _properties.setProperty("IceMX.Metrics." + name + ".Disabled", "1");
        }
        updateViews();
    }

    @Override
    public synchronized MetricsAdmin.GetMetricsViewResult getMetricsView(
            String viewName, Current current) throws UnknownMetricsView {
        MetricsAdmin.GetMetricsViewResult r = new MetricsAdmin.GetMetricsViewResult();
        MetricsViewI view = getMetricsView(viewName);
        r.timestamp = Time.currentMonotonicTimeMillis();
        if (view != null) {
            r.returnValue = view.getMetrics();
        } else {
            r.returnValue = new HashMap<>();
        }
        return r;
    }

    @Override
    public synchronized MetricsFailures[] getMapMetricsFailures(
            String viewName, String mapName, Current current) throws UnknownMetricsView {
        MetricsViewI view = getMetricsView(viewName);
        if (view != null) {
            return view.getFailures(mapName);
        }
        return new MetricsFailures[0];
    }

    @Override
    public synchronized MetricsFailures getMetricsFailures(
            String viewName, String mapName, String id, Current current) throws UnknownMetricsView {
        MetricsViewI view = getMetricsView(viewName);
        if (view != null) {
            return view.getFailures(mapName, id);
        }
        return new MetricsFailures();
    }

    public <T extends Metrics> void registerMap(String map, Class<T> cl, Runnable updater) {
        boolean updated;
        MetricsMapFactory<T> factory;
        synchronized (this) {
            factory = new MetricsMapFactory<T>(updater, cl);
            _factories.put(map, factory);
            updated = addOrUpdateMap(map, factory);
        }
        if (updated) {
            factory.update();
        }
    }

    public synchronized <S extends Metrics> void registerSubMap(
            String map, String subMap, Class<S> cl, java.lang.reflect.Field field) {
        boolean updated;
        MetricsMapFactory<?> factory;
        synchronized (this) {
            factory = _factories.get(map);
            if (factory == null) {
                return;
            }
            factory.registerSubMap(subMap, cl, field);
            removeMap(map);
            updated = addOrUpdateMap(map, factory);
        }
        if (updated) {
            factory.update();
        }
    }

    public void unregisterMap(String mapName) {
        boolean updated;
        MetricsMapFactory<?> factory;
        synchronized (this) {
            factory = _factories.remove(mapName);
            if (factory == null) {
                return;
            }
            updated = removeMap(mapName);
        }
        if (updated) {
            factory.update();
        }
    }

    public <T extends Metrics> List<MetricsMap<T>> getMaps(String mapName, Class<T> cl) {
        List<MetricsMap<T>> maps = new ArrayList<>();
        for (MetricsViewI v : _views.values()) {
            MetricsMap<T> map = v.getMap(mapName, cl);
            if (map != null) {
                maps.add(map);
            }
        }
        return maps;
    }

    public Logger getLogger() {
        return _logger;
    }

    @Override
    public void accept(Map<String, String> props) {
        for (Map.Entry<String, String> e : props.entrySet()) {
            if (e.getKey().indexOf("IceMX.") == 0) {
                // Update the metrics views using the new configuration.
                try {
                    updateViews();
                } catch (Exception ex) {
                    _logger.warning(
                        "unexpected exception while updating metrics view configuration:\n"
                            + ex.toString());
                }
                return;
            }
        }
    }

    private MetricsViewI getMetricsView(String name) throws UnknownMetricsView {
        MetricsViewI view = _views.get(name);
        if (view == null) {
            if (!_disabledViews.contains(name)) {
                throw new UnknownMetricsView();
            }
            return null;
        }
        return view;
    }

    private boolean addOrUpdateMap(String mapName, MetricsMapFactory<?> factory) {
        boolean updated = false;
        for (MetricsViewI v : _views.values()) {
            updated |= v.addOrUpdateMap(_properties, mapName, factory, _logger);
        }
        return updated;
    }

    private boolean removeMap(String mapName) {
        boolean updated = false;
        for (MetricsViewI v : _views.values()) {
            updated |= v.removeMap(mapName);
        }
        return updated;
    }

    private final Properties _properties;
    private final Logger _logger;
    private final Map<String, MetricsMapFactory<?>> _factories =
        new HashMap<>();

    private Map<String, MetricsViewI> _views = new HashMap<>();
    private final Set<String> _disabledViews = new HashSet<>();
}
